### PR TITLE
optionally bypass scatter for DistributedDataParallel

### PR DESCRIPTION
**Pull request by astooke**
Replacing pull request #22384, by making the feature optional, because some tests depended on scatter being present even for single-device.

edit:  The default behavior is the previous behavior.

### CODE REVIEW

**Comment by astooke** [comment]
My use case was, some custom data structure (looks like a tuple but quacks like a dict) I was passing into the model was not handled correctly by the scatter functionality, breaking the code.  But if using only once device in each python process, it should often be possible to run the model just on the local, un-modified inputs?

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -441,7 +449,7 @@ def forward(self, *inputs, **kwargs):
         if self.require_forward_param_sync:
             self._sync_params()
 
-        if self.device_ids:
+        if self.device_ids and (len(self.device_ids) > 1 or self.scatter_single_device):
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py]: I think #22384 failed because the input was not copied from CPU to GPU. Could you please try if the following would work?

```python

# add a internal helper function
def to(self, obj, device):
    if isinstance(obj, torch.Tensor):
        return obj.to(device)
    if isinstance(obj, tuple) and len(obj) > 0:
        return list(zip(*map(self.to, obj)))
    if isinstance(obj, list) and len(obj) > 0:
        return list(map(list, zip(*map(self.to, obj))))
    if isinstance(obj, dict) and len(obj) > 0:
        return list(map(type(obj), zip(*map(self.to, obj.items()))))
    return obj

# add this to the above line in forward function
if self.device_ids:
    if len(self.device_ids) == 1:
        inputs = self.to(inputs, self.device_ids[0])
        kwags = self.to(kwargs, self.deivce_ids[0])
        output = self.module(*inputs[0], **kwargs[0])
    else:
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
        output = self.gather(outputs, self.output_device)
else:
    ...
```

The reason for the above suggestion is that I am a little hesitate to expose a flag for a very specific use case, and it seems this can be addressed without that flag.

@gchanan Do we have any API similar to the `self.to(obj, device)` above that we can use here?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -217,6 +217,11 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is correctly used.
                          (default: ``False``)
+        scatter_single_device:  when set to ``False``, scatter functionality is
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py]: If we have to expose this flag, would it be clear if we jus call it `scatter_to_devices`, which governs scatter or not regardless the value of `device_ids`? Then the condition in forward:

```python
if self.device_ids and (len(self.device_ids) > 1 or self.scatter_single_device):
```

Becomes

```python
if self.device_ids and self.scatter_to_devices:
```

Will the above miss any use cases?

**Review by mrshenli — COMMENTED** [review]
If we can avoid exposing a new flag, I would vote for not exposing it. We already have many flags in DDP, which might look confusing. I added a suggestion on how to avoid this flag. Please have a try and see if that works for your use case.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (astooke) [reply torch/nn/parallel/distributed.py]: Thanks for the helper function, but this will also break the (possibly arbitrarily nested) data structure in my case.  I have a similar helper function intended to work for either tuples or namedtuples (and our custom namedarraytuples), which have a different construction style:

```
def buffer_to(buffer_, device=None):
    if buffer_ is None:
        return
    if isinstance(buffer_, torch.Tensor):
        return buffer_.to(device)
    elif isinstance(buffer_, np.ndarray):
        raise TypeError("Cannot move numpy array to device.")
    contents = tuple(buffer_to(b, device=device) for b in buffer_)
    if type(buffer_) is tuple:
        return contents
    return type(buffer_)(*contents)
```

Not sure if you would want to support something so specific, probably not.

But I don't understand why the DDP model should be responsible for moving inputs to the GPU when it's only running a single device, that's not consistent with the usual model, which expects data to already be on the right device.  Would prefer to bypass this intermediary processing.

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (astooke) [reply torch/nn/parallel/distributed.py]: This sounds simpler to me!  Give the user the option to manually handle the scatter even for multi-GPU.  I like it  :)

Comment (astooke) [reply torch/nn/parallel/distributed.py]: Actually, I'm not sure how that would work to use multi-GPU per process and not scatter the data inside the model.  I just don't have experience with multi-GPU per process, so I was trying to avoid changing anything to do with that.

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (mrshenli) [reply torch/nn/parallel/distributed.py]: > Not sure if you would want to support something so specific, probably not.

Right, this is too specific to fit in DDP. And I see why it the `to` helper I posted above wouldn't work.

> But I don't understand why the DDP model should be responsible for moving inputs to the GPU when it's only running a single device, that's not consistent with the usual model, which expects data to already be on the right device.

This is a good point. I think the motivation might be keep it consistent with `DataParallel` which does scatter inputs and args.

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (mrshenli) [reply torch/nn/parallel/distributed.py]: Ahh, you are right, how about this?

```python
if self.device_ids:
    if self.scatter_to_devices:
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
    if len(self.device_ids) == 1:
        output = self.module(*inputs[0], **kwargs[0])
    else:
        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
        output = self.gather(outputs, self.output_device)
else:
```

Comment (astooke) [reply torch/nn/parallel/distributed.py]: In the case of no scattering and single device, that would still execute the problematic line:
```python
output = self.module(*inputs[0], **kwargs[0])
```
which needs to be:
```python
output = self.module(*inputs, **kwargs)
```
like in the PR. :)

Comment (astooke) [reply torch/nn/parallel/distributed.py]: Could keep the PR logic but change the name to `scatter_to_device` if that's more clear?

Comment (mrshenli) [reply torch/nn/parallel/distributed.py]: Yes, let do that for now.

Comment (mrshenli) [reply torch/nn/parallel/distributed.py]: If you could please make changes to the current PR I will land it. Thanks for contributing!

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (gchanan) [reply torch/nn/parallel/distributed.py]: I'm not aware of any function like that that exists, although I don't see anything Pytorch / "to" specific about it

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (astooke) [reply torch/nn/parallel/distributed.py]: cool, done!  thanks for accepting :)

**Review by mrshenli — APPROVED** [review]

**Comment by pietern** [comment]
Thanks for the contribution! I think we should not make this conditional on `len(device_ids) == 1`.  Instead, we can make sure the inputs and kwargs are initialized identically regardless. Then the comment for the new keyword argument can be simplified and shortened and the condition in the code can be simplified to what @mrshenli suggested.

```python
if self.scatter_to_device:
    inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
else:
    inputs = [inputs]
    kwargs = [kwargs]
```

**Comment by astooke** [comment]
OK, and that block would go inside the `if self.device_ids:` block?

I'm still not sure how that would work in the case of actually having multiple devices, but using `scatter_to_device=False`.  It would make `len(inputs)==1` so that `parallel_apply()` would be called but only using the first module copy?  Kind of strange that this flag should interact with that case?

Once again, not my intended use case, so, happy to go with whatever logic you want.  :)

**Comment by mrshenli** [comment]
How about we stay with the previous suggestions, and describe the requirements clearly in the doc? I am referring to this prior one:

```python
if self.device_ids:
    if self.scatter_to_devices:
        inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
    if len(self.device_ids) == 1:
        output = self.module(*inputs[0], **kwargs[0])
    else:
        outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
        output = self.gather(outputs, self.output_device)
else:
```

The concern before was:

> In the case of no scattering and single device, that would still execute the problematic line:

Let's discuss what would be the expectation for `scatter_to_devices`. One reasonable interpretation would be, the provided inputs should always match the `devices` arg. So, if the `devices` arg is a list of a single element, the `inputs` and `kwargs` should be so as well. In that way, the following code is actually correct, and we don't need to special case `len(devices) == 1 or not`:

```python
output = self.module(*inputs[0], **kwargs[0])
```

**Comment by astooke** [comment]
Yes, I guess there is a question of whether to keep the single-GPU, non-scatter DDP interface more consistent with a) multi-GPU per process vs b) a bare model.  

My particular use case is best served with code which looks the same whether running a serial program or a parallel program with one GPU per process (i.e. each process calls the DDP-wrapped model on its own data, no scatter, only all-reduce of gradients--the least possible change).  So for me it makes sense that if you know you're not trying to scatter the inputs, to not have to put the inputs and kwargs in an extra layer of list just so it can be indexed at [0]...I'll have to go in somewhere and add some other model wrapper to do this, kind of annoying when single-GPU per process is the recommended use case.

But if you're trying to make it more like multi-GPU DDP, then yeah that makes sense that you could turn off all scattering inside the model, and assume the user has taken care of it and already organized the data per-device in a list.

**Comment by astooke** [comment]
How about:

```python
if self.device_ids:
  if self._scatter:
    inputs, kwargs = self.scatter(inputs, kwargs, self.device_ids)
  elif len(self.device_ids) == 1:
    output = self.module(*inputs, **kwargs)
  if self._scatter or len(self.device_ids) > 1:
      outputs = self.parallel_apply(self._module_copies[:len(inputs)], inputs, kwargs)
       output = self.gather(outputs, self.output_device)
else:
  output = self.module(*inputs, **kwargs)
```

And we change the flag to just be `scatter`?

**Comment by mrshenli** [comment]
I am OK with naming it `_scatter`. Regarding the code, I think it's correct and I agree that it's probably uses minimum LoC, but I feel it's a little hard to follow (maybe it's just me). I would vote for sth more verbose but easier to read and reason, e.g.:

```python
if self._scatter:
    # existing code
else:
    if len(self.device_ids) == 1:
    else:
```

I would be happy land @astooke's solution if people all agree on that.

cc @pietern

**Comment by astooke** [comment]
Waiting for a thumbs up from @pietern then I would implement @mrshenli's solution and try to squash into one commit?

**Comment by pietern** [comment]
@astooke Sorry for the delay. That sounds good to me.

We need to scatter because of the historical parallels between `DataParallel` and `DistributedDataParallel` and the support for multiple replicas in a single process. If we got rid of that we would only be left with the implicit CPU-to-GPU copy of the input tensors for GPU models. Not saying we should change DDP, but only how we got to the point we're at now :-)

**Comment by astooke** [comment]
OK sounds good!  Whew I lost this for a pit but putting it all together now.

**Comment by astooke** [comment]
OK I made the changes, squashed into one commit, rebased...hopefully good to go.

**Review by pietern — COMMENTED** [review]
Thanks for the update, @astooke!

There's still that bit of duplication that doesn't feel right... but we've already beaten this to death so I'm fine merging this and rolling with it. @mrshenli what do you think?

**Comment by astooke** [comment]
@mrshenli Just checking back before we all forget what this is about!  :)

**Comment by pietern** [comment]
@astooke Sorry about the delays here. I was looking at this again and saw we don't have tests yet! Could you add one for scatter=False with single-device and one with multi-device? Then we can ensure this doesn't regress in the future.

**Comment by mrshenli** [comment]
@astooke sorry about the delay. I agree with @pietern that it will be better to have a test cover the new code path. Thanks!

**Comment by astooke** [comment]
OK sorry about the re-delay on my end haha how do I go about adding tests?  I haven't done that before. @pietern @mrshenli

**Comment by rohan-varma** [comment]
It seems like this would be useful to simplify the code in the single-device case where one process works on a single GPU. Is there any reason we don't just do this by default and keep the scatter for multi-GPU? If it's only the tests, we can probably just refactor those tests to not rely on the scatter being called for the single-device case.
